design/one-pager-package-format-v2.md

```diff
@@ -0,0 +1,248 @@
+    source: github.com/crossplane-contrib/config-example
+    license: Apache-2.0
+    description: |
+      The Example configuration adds example resources to Crossplane.
+    provider: example
+# Required.
+spec:
+  dependsOn:
+    # Required. Specifies an OCI image containing a package dependency. This key
+    # may be either 'provider' or 'configuration'. This is sugar; in either case
+    # the package manager determines whether the depencency is really a Provider
+    # or a Configuration by unpacking it and inspecting its kind.
+  - provider: example/provider-example
+    # Required. Will be extended to support version ranges in future, but
+    # currently treated as a specific version tag.
+    version: v0.1.0
+    # Required. Specifies an OCI image containing a package dependency. This key
+    # may be either 'provider' or 'configuration'. This is sugar; in either case
+    # the package manager determines whether the depencency is really a Provider
+    # or a Configuration by unpacking it and inspecting its kind.
+  - configuration: example/some-dependency
+    # Required. Will be extended to support version ranges in future, but
+    # currently treated as a specific version tag.
+    version: v0.2.0
+  # Optional. Matching paths will be omitted from the OCI image when building a
+  # package.
+  ignore:
+    # Required. Currently only a path, relative to crossplane.yaml, is allowed.
+  - path: examples/
+```
+
+The above `Configuration` example is exhaustive - it contains all supported
+fields. It should be treated as the authoritative example of Configuration
+metadata. As with the `Provider` example above it is not a Kubernetes custom
+resource. It corresponds to a `Configuration` custom resource in the
+`packages.crossplane.io` API group. The `meta.packages.crossplane.io` file is
+authored by the provider maintainer, while the `packages.crossplane.io` custom
+resource is authored by the platform operator.
+
+## Open Question: Unpacking and Indexing Packages
+
+Two known systems consume Crossplane packages and their metadata; the Crossplane
+package manager itself and the Upbound Cloud [registry]. In each case the system
+introspects the filesystem inside the package OCI image, relying on well known
+file paths (e.g. `.registry/app.yaml`) to extract package metadata. This will be
+discouraged going forward. Instead packages will be required to supply an OCI
+image `ENTRYPOINT` that emits the metadata necessary to run or index a package.
+This 'unpack' format is currently undefined, and may influence:
+
+* Whether Crossplane can truly be unopinionated about directory structures and
+  paths within a package's OCI image.
+* How metadata such as UI annotations and icons should be associated with the
+  Crossplane resources they correspond to. This metadata is _mostly_ opaque
+  to Crossplane, but the package manager does currently merge it onto the
+  relevant Crossplane resources (e.g. CRDs) as annotations.
```

Comment:
We might at some point consider whether the "on disk" package format should be the same as the "in image" package format. Perhaps the "on disk" format would look something like this proposal, but the package CLI tooling would compile it all into a single stream of YAML at build time. The package would therefore contain only a single file containing a stream of YAML, and its `ENTRYPOINT` would be roughly `cat big-stream-of.yaml`.

One approach here might be to frame on disk metadata like UI schemas, icons, etc as "what resource should I annotate, and how?". Rather than storing that metadata in `.crossplane` we could introduce a new type that the package tooling could understand, like:

```yaml
apiVersion: meta.packages.crossplane.io/v1alpha1
kind: Extensions
metadata:
  name: rdsinstances.database.aws.crossplane.io
spec:
  # This Extension extends the RDSInstance CRD.
  extendsRef:
    apiVersion: apiextensions.k8s.io/v1beta1
    kind: CustomResourceDefinition
    name: rdsinstances.database.aws.crossplane.io
  # It adds the following annotations. Perhaps this should be an overlay,
  # but it could be useful to support reading values from files.
  annotations:
  - key: registry.upbound.io/icon
     valueFromFile: icon.svg
  - key: registry.upbound.io/ui-schema
    value: |-
      ---
      configSections:
      - title: Configuration
        description: Enter information specific to the configuration you wish to create.
        items:
        - name: dbReplicas
          controlType: singleInput
          type: integer
          path: ".spec.dbReplicas"
          title: DB Replicas
          description: The number of DB Replicas
          default: 1
          validation:
          - minimum: 1
          - maximum: 3
```

Reply:
If we invoke the `ENTRYPOINT` to get the steam of yaml how would a registry get binary icon data? It doesn't seem like streaming that to STDOUT would work very well. Would package containers have to provide both an ENTRYPOINT and icon files at specific paths in the container image?

Reply:
> If the "in image" contents are just a static stream of yaml that was assembled at build time, it (may) be harder to be able to take into account install time variables or runtime environment information

@jbw976  In retrospect, I agree - I'm now convinced there's value in this and have updated the design to match. I don't think we should be too open ended here though; I've currently proposed a single `--context` flag that may only alter how the YAML stream is annotated.

> If we invoke the ENTRYPOINT to get the steam of yaml how would a registry get binary icon data?

I believe the current thinking is that the binary icon data should be emitted as annotations on the relevant Crossplane document (e.g. CRD, XRD, etc) in the stream. I know there were some concerns about the size of such data in an annotation. I'm hoping we could control this with the `--context` flag - e.g. `--context=index` would include the annotations we need for indexing (including image data), while `--context=unpack` would not.